### PR TITLE
Lv2: Improve plugin description

### PR DESCRIPTION
**Pull request by JohannesLorenz**
What this does:

1. Add a description tooltip to the Lv2 instrument selection
2. Extend the effect description widget in the EffectSelectDialog

This PR is ready for code and testing review.

### CODE REVIEW

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 1]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+	auto container = new QWidget(parent);
+	auto l = new QHBoxLayout(container);
 	l->setContentsMargins(0, 0, 0, 0);
 	l->setSpacing(0);
 
-	auto maker_label = new QLabel(maker);
-	maker_label->setText(QWidget::tr("Maker: "));
-	maker_label->setAlignment(Qt::AlignTop);
+	auto leftLabel = new QLabel(container);
+	leftLabel->setText(QWidget::tr(left));
```
Comment (Rossmaxx) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:82]: I believe you are missing quotes, or is it this way?

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 2]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
-		qStringFromPluginNode(plug, lilv_plugin_get_author_name));
-	maker_content->setWordWrap(true);
+	auto rightLabel = new QLabel(container);
+	if(right.startsWith("http") && !right.contains(' ') && !right.contains('\n'))
+	{
+		right = QString("<a href=\"%1\">%2</a>").arg(right, right);
+		rightLabel->setTextInteractionFlags(rightLabel->textInteractionFlags()
+			| Qt::LinksAccessibleByMouse | Qt::LinksAccessibleByKeyboard);
+		rightLabel->setTextFormat(Qt::RichText);
+		QObject::connect(rightLabel, &QLabel::linkActivated, [](const QString & link){
```
Comment (Rossmaxx) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:92]: Can we have an actual function instead of a lambda, it's less clear this way? Feel free to ignore tho.

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 3]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+{
+	auto lbl = new QLabel(parent);
+	if(right.isEmpty()) { return; }
+	if(right.startsWith("http") && !right.contains(' ') && !right.contains('\n'))
+	{
+		right = QString("<a href=\"%1\">%2</a>").arg(right, right);
+		lbl->setTextInteractionFlags(lbl->textInteractionFlags()
+			| Qt::LinksAccessibleByMouse | Qt::LinksAccessibleByKeyboard);
+		lbl->setTextFormat(Qt::RichText);
+		QObject::connect(lbl, &QLabel::linkActivated, [](const QString & link){
```
Comment (Rossmaxx) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:115]: We use the same lambda here, so making it a function call will add benefit.

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 4]
```diff
@@ -144,8 +220,21 @@ QString Lv2SubPluginFeatures::displayName(
 QString Lv2SubPluginFeatures::description(
 	const Plugin::Descriptor::SubPluginFeatures::Key &k) const
 {
-	(void)k;
-	return QString::fromUtf8("description not implemented yet"); // TODO
+	auto mgr = Engine::getLv2Manager();
+	const LilvPlugin *plug = mgr->getPlugin(k.attributes["uri"]);
```
Comment (Rossmaxx) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:224]:
```suggestion
	const LilvPlugin* plug = mgr->getPlugin(k.attributes["uri"]);
```

Minor style suggestion.

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 5]
```diff
@@ -154,8 +243,8 @@ QString Lv2SubPluginFeatures::description(
 const PixmapLoader *Lv2SubPluginFeatures::logo(
 	const Plugin::Descriptor::SubPluginFeatures::Key &k) const
 {
-	(void)k; // TODO
-	return nullptr;
+	(void)k;
+	return nullptr; // Lv2 currently does not support this
```
Comment (Rossmaxx) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:247]: Why the changes in comment? Also the C style cast seems to have no place here. The cast looks fine to remove, or is it to suppress unused warnings?

**Review by Rossmaxx — COMMENTED** [review]
Minor review.

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 5 above)
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:247]: The change is done because I previously thought we could implement this later. So I now wanted to implement this function, but found out Lv2 does not support it. Thus, I updated the comment.

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 4 above)
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:224]: We do not have a style convention for this stuff. I fear if I change this, either someone will tell me to change it consistent in the whole file, or will tell me that I should change it back :smiley:

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 1 above)
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:82]: Quotes where exactly/for what?

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 2 above)
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:92]: I usually start making functions if two pieces of code have at least 2 statements (2 lines) in common, but in this case, it is just one line. IMO it looks readable that way.

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 3 above)
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:115]: Same argumentation here. Putting one line into a common functions seems overkill (to me).

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 4 above)
Comment (Rossmaxx) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:224]: Well, i do get confused about this often, so that's why i change, and ask to change in review. I thought this was a part of the convention. Sorry for the confusion.

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 1 above)
Comment (Rossmaxx) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:82]: Within the tr()

Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:82]: You mean `tr("left")`? I want it to really use the text which had been passed to the function, not the constant string `"left"`.

Comment (Rossmaxx) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:82]: Ahh, why did i leave such a dum review?

Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:82]: All good, thanks for trying!

**Review by sakertooth — APPROVED** [review]
Seems like I reviewed this PR for style awhile ago, just never sent it for some reason. LGTM otherwise, though someone who is more familiar with Lv2 would probably be a more credible approval than mine.

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 6]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+	auto leftLabel = new QLabel(container);
+	leftLabel->setText(QWidget::tr(left));
+	leftLabel->setAlignment(Qt::AlignTop);
 
-	auto maker_content = new QLabel(maker);
-	maker_content->setText(
-		qStringFromPluginNode(plug, lilv_plugin_get_author_name));
-	maker_content->setWordWrap(true);
+	auto rightLabel = new QLabel(container);
+	if (right.startsWith("http") && !right.contains(' ') && !right.contains('\n'))
```
Comment (messmerd) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:86]: I'm sure there are better ways to do URL detection, but practically speaking, this should work

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 7]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+{
+	assert(plug);
+	AutoLilvNode valueUriNode{Engine::getLv2Manager()->uri(valueUri)};
+	return AutoLilvNodes{lilv_plugin_get_value(plug, valueUriNode.get())};
+}
+
+
+
+
+void Lv2SubPluginFeatures::fillDescriptionWidget(QWidget *parent,
```
Comment (messmerd) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:134]: Some minor style issues here and there with the pointers

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 8]
```diff
@@ -144,8 +220,21 @@ QString Lv2SubPluginFeatures::displayName(
...
+		AutoLilvNode rdfs_comment{mgr->uri(LILV_NS_RDFS "comment")};
+		AutoLilvNodes comments{lilv_plugin_get_value(plug, rdfs_comment.get())};
+		if (comments)
+		{
+			result += "\n\n";
+			result += lilv_node_as_string(lilv_nodes_get_first(comments.get()));
+		}
+		return result.trimmed();
+	}
+	return QString::fromUtf8("failed to load description");
```
Comment (messmerd) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:237]:
```suggestion
	return QObject::tr("failed to load description");
```

**Review by messmerd — COMMENTED** [review]
I'll test it later

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 9]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
-	AutoLilvNodes extensions(lilv_plugin_get_extension_data(plug));
-	(void)extensions;
-	// possibly TODO: version, project, plugin type, number of channels
+void addLabel(QWidget* parent, const char* left, QString right)
+{
+	auto lbl = new QLabel(parent);
+	if (right.isEmpty()) { return; }
+	if (right.startsWith("http") && !right.contains(' ') && !right.contains('\n'))
+	{
+		right = QString("<a href=\"%1\">%2</a>").arg(right, right);
```
Comment (DomClark) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:111]: Should be able to repeat `%1` for multiple substitutions of the same value:
```suggestion
		right = QString("<a href=\"%1\">%1</a>").arg(right);
```

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 10]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+		lbl->setTextInteractionFlags(lbl->textInteractionFlags()
+			| Qt::LinksAccessibleByMouse | Qt::LinksAccessibleByKeyboard);
+		lbl->setTextFormat(Qt::RichText);
+		QObject::connect(lbl, &QLabel::linkActivated, [](const QString & link){
+			QDesktopServices::openUrl(link);});
```
Comment (DomClark) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:116]: `QLabel` has this functionality built in:
```suggestion
		lbl->setOpenExternalLinks(true);
```

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 11]
```diff
@@ -67,54 +69,128 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+			descLabel->setText("<i>" + description.trimmed() + "</i>");
+			descLabel->setWordWrap(true);
+		}
+	}
+
+	addLabel(parent, "URI: ", lilv_node_as_uri(lilv_plugin_get_uri(plug)));
+	addHbox(parent, "Project: ",
+		qStringFromPluginNode(plug, lilv_plugin_get_project));
+	addHbox(parent, "Maker: ",
+		qStringFromPluginNode(plug, lilv_plugin_get_author_name));
```
[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:167]: They should already be, since `addLabel` and `addHbox` use `QWidget::tr()`. What do you mean?

[the remark this answers is not in the document]
Comment (DomClark) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:167]: The tool that extracts strings to be translated, `lupdate`, looks for string literals inside calls to `tr`. If the string is passed indirectly, through some intermediate function or variable, then it won't be detected. As a result, the translation files won't contain an entry for it, so `tr` will return it unchanged.

[the remark this answers is not in the document]
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:167]: Ouch! OK, thanks, I guess this is fixed by moving the `tr` here and then passing the `tr` result to the `add...` functions?

[the remark this answers is not in the document]
Comment (DomClark) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:167]: That's right.

File: src/core/lv2/Lv2SubPluginFeatures.cpp (hunk 6 above)
Comment (JohannesLorenz) [reply src/core/lv2/Lv2SubPluginFeatures.cpp:86]: I tried to use [QUrl::isValid()](https://doc.qt.io/qt-6/qurl.html#isValid), but it recognizes "Calf Studio Gear" as a valid URL :smiley: For now, I will probably just keep it like that.

**Comment by JohannesLorenz** [comment]
All review comments fixed! Thanks for the 4 reviews, more than I could have asked :smiley: 

I keep this open another week if anyone wants to check the rework or if anyone wants to test, then I will merge.

File: src/core/lv2/Lv2SubPluginFeatures.cpp [hunk 12]
```diff
@@ -67,61 +69,135 @@ Lv2SubPluginFeatures::Lv2SubPluginFeatures(Plugin::Type type) :
...
+			? lilv_node_as_uri(lilv_nodes_get_first(licenses.get()))
+			: "<unknown>");
+	}
+	{
+		const LilvNode* libraryUriNode = lilv_plugin_get_bundle_uri(plug);
+		const char* libraryUri = lilv_node_as_uri(libraryUriNode);
+		auto filename = std::unique_ptr<char, void(*)(char*)>(
+			lilv_file_uri_parse(libraryUri, nullptr),
+			[](char* ptr) { lilv_free(ptr); }
+		);
```
Comment (sakertooth) [inline src/core/lv2/Lv2SubPluginFeatures.cpp:188]: Would this work?
```suggestion
		auto fileUri = lilv_file_uri_parse(libraryUri, nullptr);
		auto filename = std::unique_ptr<char, void (*)(void*)>(fileUri, &lilv_free);
```

**Comment by messmerd** [comment]
I think it would look better if the description came first followed by the URL

**Comment by JohannesLorenz** [comment]
> I think it would look better if the description came first followed by the URL

It seems to be already like this, both from the code (description is `pluginGetValues(plug, LILV_NS_RDFS "comment")`) and from this picture?

grafik

**Comment by messmerd** [comment]
The plugin browser looks fine, it's the tooltips when you hover over the instrument in the Instrument Plugins sidebar that the URL comes first followed by what seems to be the description or the name. The Black Pearl Drumkit exhibits this.
